internal/provider/data_source_gitlab_project_membership.go

```diff
@@ -0,0 +1,188 @@
+						},
+						"access_level": {
+							Description: "The level of access to the group.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"expires_at": {
+							Description: "Expiration date for the group membership.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+})
+
+func dataSourceGitlabProjectMembershipRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	var project *gitlab.Project
+	var err error
+
+	log.Printf("[INFO] Reading Gitlab project")
+
+	if projectID, ok := d.GetOk("project_id"); ok {
+		// Get project by id
+		project, _, err = client.Projects.GetProject(projectID.(int), nil, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	} else if fullPath, ok := d.GetOk("full_path"); ok {
+		// Get project by full path
+		project, _, err = client.Projects.GetProject(fullPath.(string), nil)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	} else {
+		return diag.Errorf("one and only one of project_id or full_path must be set")
+	}
+
+	log.Printf("[INFO] Reading Gitlab project memberships")
+
+	// Get project memberships
+	listOptions := &gitlab.ListProjectMembersOptions{
+		ListOptions: gitlab.ListOptions{
+			PerPage: 20,
+			Page:    1,
+		},
+	}
+
+	listMembers := client.ProjectMembers.ListProjectMembers
+	if inherited, ok := d.GetOk("inherited"); ok && inherited.(bool) {
+		listMembers = client.ProjectMembers.ListAllProjectMembers
+	}
+
+	var allPMs []*gitlab.ProjectMember
+	for {
+		pms, resp, err := listMembers(project.ID, listOptions, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		allPMs = append(allPMs, pms...)
+
+		if resp.NextPage == 0 {
+			break
+		}
+		listOptions.Page = resp.NextPage
+	}
+
+	d.Set("project_id", project.ID)
+	d.Set("full_path", project.PathWithNamespace)
+
+	if err := d.Set("members", flattenGitlabProjectMembers(d, allPMs)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var optionsHash strings.Builder
+	optionsHash.WriteString(strconv.Itoa(project.ID))
+
+	if data, ok := d.GetOk("access_level"); ok {
+		optionsHash.WriteString(data.(string))
+	}
+
+	id := schema.HashString(optionsHash.String())
+	d.SetId(fmt.Sprintf("%d", id))
+
+	return nil
+}
+
+func flattenGitlabProjectMembers(d *schema.ResourceData, members []*gitlab.ProjectMember) []interface{} {
+	membersList := make([]interface{}, 0, len(members))
+	for _, member := range members {
+		values := map[string]interface{}{
+			"id":           member.ID,
+			"username":     member.Username,
+			"name":         member.Name,
+			"state":        member.State,
+			"avatar_url":   member.AvatarURL,
+			"web_url":      member.WebURL,
+			"access_level": accessLevelValueToName[gitlab.AccessLevelValue(member.AccessLevel)],
+		}
+
+		if member.ExpiresAt != nil {
+			values["expires_at"] = member.ExpiresAt.String()
```

Comment:
```suggestion
			values["expires_at"] = member.ExpiresAt.Format(time.RFC3339)
```

(See docs of time.String() for why it should never be used outside of debugging)

Reply:
Should I also update the group membership with this as well?

Reply:
Also, very interesting, I didn't know!

Reply:
Let's keep the other resource as-is, since it might be a breaking change there. 😉

Reply:
Actually, `members.ExpiresAt` is a `go-gitlab.ISOTime` type, which has the following `String` implementation:
```go
func (t ISOTime) String() string {
    return time.Time(t).Format(iso8601)
}
```

So it is stable, but with the wrong format :-)

Reply:
Interesting! `gitlab.ISOTime` is a type alias of `time.Time`. Looking at the API docs - https://docs.gitlab.com/ee/api/members.html - it includes precision down to the second, but `iso8601` is just a date without time. So I would still recommend my suggestion.